### PR TITLE
Fix crashes for RealmAny on ARM 32 bit architectures

### LINKED ISSUE #7626: RealmList<RealmAny> on Android ARM devices works incorrectly
#### Expected results
List of RealmAny objects

#### Actual Results
List of "null" objects

#### Steps & Code to Reproduce
```
package com.example.testapp

import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import io.realm.*
import io.realm.annotations.PrimaryKey
import java.io.File
import kotlin.random.Random

open class FirstInnerTestEntity(
    @PrimaryKey
    var id: Long = 0
) : RealmObject()

open class SecondInnerTestEntity(
    @PrimaryKey
    var id: Long = 0
) : RealmObject()

open class TestEntity(
    @PrimaryKey
    var id: Long = 0,
    var items: RealmList<FirstInnerTestEntity>? = null
) : RealmObject()

open class WithRealmAnyTestEntity(
    @PrimaryKey
    var id: Long = 0,
    var items: RealmList<RealmAny>? = null
) : RealmObject()

class MainActivity : AppCompatActivity() {

    private lateinit var realm: Realm

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)


        initDb()

        saveEntities()

        getEntities(FirstInnerTestEntity::class.java)
        getEntities(SecondInnerTestEntity::class.java)
        getEntities(TestEntity::class.java)
        getEntities(WithRealmAnyTestEntity::class.java)

        setContentView(R.layout.activity_main)
    }

    private fun initDb() {
        Realm.init(applicationContext)

        val directory =
            File(applicationContext.filesDir.absolutePath + File.separator + "app_db")

        val config = RealmConfiguration.Builder()
            .directory(directory)
            .deleteRealmIfMigrationNeeded()
            .allowWritesOnUiThread(true)
            .build()

        Realm.setDefaultConfiguration(config)

        realm = Realm.getDefaultInstance()
    }

    private fun <C: RealmModel> getEntities(clazz: Class<C>): List<C> {
        val results = realm.where(clazz).findAll()
        return realm.copyFromRealm(results)
    }

    private fun saveEntities() {
        val entities = (0..100).realmMap {
            TestEntity(
                id = it.toLong(),
                items = (0..10).realmMap {
                    FirstInnerTestEntity(
                        Random.nextLong()
                    )
                }
            )
        }
        val entities2 = (0..100).realmMap {
            WithRealmAnyTestEntity(
                id = it.toLong(),
                items = (0..10).realmMap {
                    RealmAny.valueOf(
                        when (Random.nextBoolean()) {
                            true -> FirstInnerTestEntity(
                                Random.nextLong()
                            )
                            false -> SecondInnerTestEntity(
                                Random.nextLong()
                            )
                        }
                    )
                }
            )
        }

        realm.executeTransaction {
            realm.insertOrUpdate(entities)
            realm.insertOrUpdate(entities2)
        }
    }

    inline fun <T, R> Iterable<T>.realmMap(transform: (T) -> R): RealmList<R> {
        return mapTo(RealmList(), transform)
    }
}
```
**on realm-java 10.8.1 (realm-core 11.6.0)** this sample crashing with next stacktrace:

```
<unknown> 0x0000000091680608
Java_io_realm_internal_core_NativeRealmAny_nativeGetType 0x00000000916c4aac
int io.realm.internal.core.NativeRealmAny.nativeGetType(long) 0x0000000091a57c08
art_quick_invoke_stub_internal 0x00000000ac583376
art_quick_invoke_static_stub 0x00000000ac5885ee
art::ArtMethod::Invoke(art::Thread*, unsigned int*, unsigned int, art::JValue*, char const*) 0x00000000ac2300e2

       *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
Build fingerprint: 'motorola/pettyl/pettyl:8.1.0/OPG28.54-53-8/55beb:user/release-keys'
Revision: 'p300'
ABI: 'arm'
pid: 19549, tid: 19549, name: example.testapp  >>> com.example.testapp <<<
signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x8b
Cause: null pointer dereference
    r0 00000000  r1 0000007b  r2 0000007b  r3 00000000
    r4 bed1c6b8  r5 00000000  r6 00000000  r7 bed1c698
    r8 00000000  r9 0000007b  sl 00000000  fp bed1c76c
    ip 975c2a99  sp bed1c680  lr 9757e609  pc 9757e608  cpsr 40070030
backtrace:
    #00 pc 001fe608  /data/app/com.example.testapp-4jVqw7f-kkz-6UcIiIozOQ==/lib/arm/librealm-jni.so
    realm/realm-core#1 pc 00242aa9  /data/app/com.example.testapp-4jVqw7f-kkz-6UcIiIozOQ==/lib/arm/librealm-jni.so (Java_io_realm_internal_core_NativeRealmAny_nativeGetType+16)
    realm/realm-core#2 pc 00021f5f  /data/app/com.example.testapp-4jVqw7f-kkz-6UcIiIozOQ==/oat/arm/base.odex (offset 0x20000)
```

**on realm-java realm-java 10.9.0 (realm-core 11.7.0)**  getEntities(WithRealmAnyTestEntity::class.java)  returns list of "null" 
145781083-90383218-94dc-4acb-b568-4aaa493eb622

**Can be reproduced on ARM devices:**
SM-A105FN (API 29)
Nokia 1 (API 27)
Motorolla G5S (API 27)
Redmi 6A (API 27)
moto e5 play (API 27)
SM-G532G (API 23)
Nexus 5 (API 23)
LG-AS110 (API 23)
Nexus 7 (API 21)

#### Core version
realm-java 10.9.0 (realm-core 11.7.0)

### PR DESCRIPTION
**Pull request by rorbech**
Closes #7626

### CODE REVIEW

**Comment by rorbech** [comment]
This fix isn't verified by CI, but I have executed our full test suite as 32-bit tests on a arm devices. Dump of important output is: 
```
$ ./gradlew connectedObjectServerDebugAndroidTest

[...]

[CXX5202] This app only has 32-bit [armeabi-v7a] native libraries. Beginning August 1, 2019 Google Play store requires that all apps that include native libraries must provide 64-bit versions. For more information, visit https://g.co/64-bit-requirement

> Task :kotlin-extensions:connectedObjectServerDebugAndroidTest
Starting 61 tests on SM-G973F - 11

[...]

> Task :realm-library:connectedObjectServerDebugAndroidTest
Starting 8163 tests on SM-G973F - 11

[...]

BUILD SUCCESSFUL in 31m 54s
117 actionable tasks: 7 executed, 110 up-to-date
````

File: realm/realm-library/src/androidTest/kotlin/io/realm/RealmAnyCollectionTests.kt [hunk 1]
```diff
@@ -58,6 +61,20 @@ class RealmAnyCollectionTests {
         realm.close()
     }
 
+    @Test
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/kotlin/io/realm/RealmAnyCollectionTests.kt:64]:
```suggestion
    // Test for https://github.com/realm/realm-java/issues/7626
    @Test
```

**Review by cmelchior — APPROVED** [review]
Great find 🥇 ... but would still be interesting to understand the root cause.

**Comment by rorbech** [comment]
> Great find 🥇 ... but would still be interesting to understand the root cause.

Guess the pointer size that can differ based on compiler/platform, etc, so forcing it into `jlong` (int64_t) ensures that it isn't truncated. But if the platform somehow only used smaller types you would think that it could also fit inside the jlong passed to java. Maybe something skew issue around the `va_list` convention as it seemed like the erroneous pointer was pointing to memory containing `(J)V` which is the signature argument to the Java call.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,17 @@
+## 10.9.1 (YYYY-MM-DD)
+
+### Enhancements
+* None.
+
+### Fixed
+* Crashes when using `RealmAny` inside `RealmList`. (Issue [#7626](https://github.com/realm/realm-java/issues/7626))
```
Comment (clementetb) [inline CHANGELOG.md:7]: Shall we point that the affected platform is ARM32?

**Review by clementetb — APPROVED** [review]
Great find ⭐️!

**Comment by rorbech** [comment]
The cast is needed due to https://en.cppreference.com/w/cpp/language/reinterpret_cast point 2 and 3 (thanks James)

**Pull request merged**